core/container.go

```diff
@@ -1468,118 +1510,130 @@ func (container *Container) Publish(
 	}
 
-	exportOpts.Output = func(map[string]string) (io.WriteCloser, error) {
-		return out, nil
-	}
-	return host.Export(ctx, exportOpts, bkClient, solveOpts, solveCh, func(ctx context.Context, gw bkgw.Client) (*bkgw.Result, error) {
-		return container.export(ctx, gw, platformVariants)
+	_, err = WithServices(ctx, bk, services, func() (map[string]string, error) {
+		return bk.ExportContainerImage(ctx, inputByPlatform, dest, opts)
 	})
+	return err
 }
 
-const OCIStoreName = "dagger-oci"
-
-var importCache = newCacheMap[uint64, *Container]()
+var importCache = newCacheMap[uint64, *specs.Descriptor]()
 
 func (container *Container) Import(
 	ctx context.Context,
-	gw bkgw.Client,
-	host *Host,
-	source FileID,
-	tag string,
-	store content.Store,
-) (*Container, error) {
-	return importCache.GetOrInitialize(
-		cacheKey(container, source, tag),
-		func() (*Container, error) {
-			return container.importUncached(ctx, gw, host, source, tag, store)
-		},
-	)
-}
-
-func (container *Container) importUncached(
-	ctx context.Context,
-	gw bkgw.Client,
+	bk *buildkit.Client,
 	host *Host,
 	source FileID,
 	tag string,
 	store content.Store,
+	lm *leaseutil.Manager,
 ) (*Container, error) {
 	file, err := source.ToFile()
 	if err != nil {
 		return nil, err
 	}
 
-	src, err := file.Open(ctx, host, gw)
-	if err != nil {
-		return nil, err
-	}
+	var release func(context.Context) error
+	loadManifest := func() (*specs.Descriptor, error) {
+		src, err := file.Open(ctx, host, bk)
+		if err != nil {
+			return nil, err
+		}
 
-	defer src.Close()
+		defer src.Close()
 
-	container = container.Clone()
+		container = container.Clone()
 
-	stream := archive.NewImageImportStream(src, "")
+		// override outer ctx with release ctx and set release
+		ctx, release, err = leaseutil.WithLease(ctx, lm, leaseutil.MakeTemporary)
+		if err != nil {
+			return nil, err
+		}
+
+		stream := archive.NewImageImportStream(src, "")
+
+		desc, err := stream.Import(ctx, store)
+		if err != nil {
+			return nil, fmt.Errorf("image archive import: %w", err)
+		}
+
+		return resolveIndex(ctx, store, desc, container.Platform, tag)
+	}
 
-	desc, err := stream.Import(ctx, store)
+	// TODO: seems ineffecient to recompute for each platform, but do need to get platform-specific manifest stil..
+	key := cacheKey(file, tag, container.Platform)
+
+	manifestDesc, err := importCache.GetOrInitialize(key, loadManifest)
 	if err != nil {
-		return nil, fmt.Errorf("image archive import: %w", err)
+		return nil, err
 	}
 
-	manifestDesc, err := resolveIndex(ctx, store, desc, container.Platform, tag)
+	_, err = store.Info(ctx, manifestDesc.Digest)
 	if err != nil {
-		return nil, fmt.Errorf("image archive resolve index: %w", err)
+		// TODO(vito): loadManifest again, to be durable to buildctl prune. but I
+		// can't reproduce this at the moment since it doesn't seem to be pruned.
+		return nil, fmt.Errorf("manifest pruned?: %w", err)
```

Comment:
@vito is this still worth keeping in?

Reply:
Investigating; I think it's a real issue, but I'm having trouble repro'ing it with our existing Import tests. Going to change them to use `apko` to build an image instead of relying on Export + Host APIs.

Reply:
Ok, so I'm able to repro this but only after _also_ changing the cache keys to use stable digests. I think right now the caching is just not working, so this secondary edge case isn't able to happen. Basically two wrongs are making a right.

[repro diff + logs](https://gist.github.com/vito/a3930b759d80336d07f3caf8c34d8af1#file-import-repro-result-txt)

`main` is just as broken so I think this can either be removed or left as a TODO if I don't push a commit in time.